tools/cabana/streams/routes.cc

```diff
@@ -79,45 +96,78 @@ void RoutesDialog::fetchRoutes() {
+  if(isPreservedTabSelected()) {
+    QString url = QString("%1/v1/devices/%2/routes/preserved")
+                      .arg(CommaApi::BASE_URL)
+                      .arg(dongle_id);
+    http->sendRequest(url);
+  } else {
+    QDateTime current = QDateTime::currentDateTime();
+    QString url = QString("%1/v1/devices/%2/routes_segments?start=%3&end=%4")
+                      .arg(CommaApi::BASE_URL)
+                      .arg(dongle_id)
+                      .arg(current.addDays(-(period_selector_->currentData().toInt())).toMSecsSinceEpoch())
+                      .arg(current.toMSecsSinceEpoch());
+    http->sendRequest(url);
+  }
 }
 
 void RoutesDialog::parseRouteList(const QString &json, bool success, QNetworkReply::NetworkError err) {
   if (success) {
     for (const QJsonValue &route : QJsonDocument::fromJson(json.toUtf8()).array()) {
-      uint64_t start_time = route["start_time_utc_millis"].toDouble();
-      uint64_t end_time = route["end_time_utc_millis"].toDouble();
-      auto datetime = QDateTime::fromMSecsSinceEpoch(start_time);
-      auto item = new QListWidgetItem(QString("%1    %2min").arg(datetime.toString()).arg((end_time - start_time) / (1000 * 60)));
+      uint64_t to;
+      QDateTime from;
+
+      if(isPreservedTabSelected()) {
+        QString start_time = route["start_time"].toString();
+        QString end_time = route["end_time"].toString();
+
+        from = QDateTime::fromString(start_time, Qt::ISODateWithMs);
+        to = QDateTime::fromString(end_time, Qt::ISODateWithMs).msecsTo(from);
+      } else {
+        uint64_t start_time = route["start_time_utc_millis"].toDouble();
+        uint64_t end_time = route["end_time_utc_millis"].toDouble();
+        from = QDateTime::fromMSecsSinceEpoch(start_time);
+        to = (end_time - start_time);
+      }
+
+      QListWidgetItem *item = new QListWidgetItem(QString("%1    %2min").arg(from.toString()).arg(to / (1000 * 60)));
       item->setData(Qt::UserRole, route["fullname"].toString());
-      route_list_->addItem(item);
+      currentRoutesList()->addItem(item);
     }
     // Select first route if available
-    if (route_list_->count() > 0) route_list_->setCurrentRow(0);
+    if (currentRoutesList()->count() > 0) currentRoutesList()->setCurrentRow(0);
   } else {
     QMessageBox::warning(this, tr("Error"), tr("Failed to fetch routes. Check your network connection."));
     reject();
   }
-  route_list_->setEmptyText(tr("No items"));
+  currentRoutesList()->setEmptyText(tr("No items"));
   sender()->deleteLater();
 }
 
 void RoutesDialog::accept() {
-  if (auto current_item = route_list_->currentItem()) {
+  if (auto current_item = currentRoutesList()->currentItem()) {
     route_ = current_item->data(Qt::UserRole).toString();
   }
   QDialog::accept();
 }
+
+RouteListWidget* RoutesDialog::currentRoutesList() {
+  if(routes_type_selector_->currentIndex() == 1) {
+    return preserved_route_list_;
+  }
+
+  return route_list_;
+}
```

Comment:
Using a ternary operator here can make the code both more concise and readable.:
`return routes_type_selector_->currentIndex() == 0 ? route_list_ : preserved_route_list_;`